roachtest: version/mixed/nodes=5 failed

roachtest.version/mixed/nodes=5 [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3366874&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3366874&tab=artifacts#/version/mixed/nodes=5) on master @ [8cae60f603ccc4d83137167b3b31cab09be9d41a](https://github.com/cockroachdb/cockroach/commits/8cae60f603ccc4d83137167b3b31cab09be9d41a):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/version/mixed/nodes=5/run_1
	monitor.go:128,version.go:211,version.go:224,test_runner.go:777: monitor failure: unexpected node event: 4: dead (exit status 134)
		(1) attached stack trace
		  -- stack trace:
		  | main.(*monitorImpl).WaitE
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/monitor.go:116
		  | main.(*monitorImpl).Wait
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/monitor.go:124
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerVersion.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/version.go:211
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerVersion.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/version.go:224
		  | main.(*testRunner).runTest.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:777
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		Wraps: (2) monitor failure
		Wraps: (3) unexpected node event: 4: dead (exit status 134)
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *errors.errorString

	cluster.go:1249,context.go:89,cluster.go:1237,test_runner.go:866: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-3366874-1630046181-38-n6cpu4 --oneshot --ignore-empty-nodes: exit status 1 4: dead (exit status 134)
		6: skipped
		5: 17200
		1: 21767
		3: 22362
		2: 20320
		Error: UNCLASSIFIED_PROBLEM: 4: dead (exit status 134)
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1173
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:281
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:856
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:960
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:897
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:2107
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:225
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		Wraps: (3) 4: dead (exit status 134)
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError
```
<details><summary>Reproduce</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*version/mixed/nodes=5.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

```
I210827 09:23:53.598678 1 util/log/flags.go:180  [-] 1  stderr capture started
panic: Payload.Type called on a payload with an unknown details type: <nil>

goroutine 2254823 [running]:
panic(0x47c9220, 0xc0008cd4b0)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc007da7980 sp=0xc007da78b8 pc=0x48b725
github.com/cockroachdb/cockroach/pkg/jobs/jobspb.DetailsType(0x0, 0x0, 0x14)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobspb/wrap.go:89 +0x385 fp=0xc007da79e0 sp=0xc007da7980 pc=0x1f1bfc5
github.com/cockroachdb/cockroach/pkg/jobs/jobspb.(*Payload).Type(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobspb/wrap.go:56
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).createResumer(0xc000615a00, 0xc006ed2140, 0xc0002d0000, 0x0, 0x0, 0x0, 0x4c3b57c)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1146 +0x91 fp=0xc007da7b98 sp=0xc007da79e0 pc=0x23cda71
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob(0xc000615a00, 0x5a99a00, 0xc001760ac0, 0x98c9510c6658001, 0x5a43c40, 0xc000e1ec60, 0xc001465400, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:209 +0x6b5 fp=0xc007da7f10 sp=0xc007da7b98 pc=0x23bb355
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeClaimedJobs.func3(0xc006780d80, 0xc000615a00, 0x5a99a00, 0xc001760ac0, 0x5a43c40, 0xc000e1ec60, 0x98c9510c6658001)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:123 +0xaa fp=0xc007da7fa8 sp=0xc007da7f10 pc=0x23d81aa
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc007da7fb0 sp=0xc007da7fa8 pc=0x4c4b41
created by github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeClaimedJobs
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:121 +0x1aa

```

cc @ajwerner

--- a/pkg/sql/sqlstats/persistedsqlstats/BUILD.bazel
+++ b/pkg/sql/sqlstats/persistedsqlstats/BUILD.bazel
@@ -58,6 +58,7 @@ go_test(
         "compaction_test.go",
         "controller_test.go",
         "flush_test.go",
+        "job_monitor_test.go",
         "main_test.go",
         "reader_test.go",
         "scheduled_sql_stats_compaction_test.go",
@@ -65,6 +66,7 @@ go_test(
     deps = [
         ":persistedsqlstats",
         "//pkg/base",
+        "//pkg/clusterversion",
         "//pkg/jobs",
         "//pkg/jobs/jobspb",
         "//pkg/jobs/jobstest",

--- a/pkg/sql/sqlstats/persistedsqlstats/job_monitor_test.go
+++ b/pkg/sql/sqlstats/persistedsqlstats/job_monitor_test.go
@@ -1,0 +1,63 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package persistedsqlstats_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/server"
+	"github.com/cockroachdb/cockroach/pkg/sql/tests"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+)
+
+func TestVersionGating(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	params, _ := tests.CreateTestServerParams()
+	params.Knobs.Server = &server.TestingKnobs{
+		DisableAutomaticVersionUpgrade: 1,
+		BinaryVersionOverride:          clusterversion.ByKey(clusterversion.SQLStatsCompactionScheduledJob - 1),
+	}
+	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
+		ServerArgs: params,
+	})
+
+	defer tc.Stopper().Stop(ctx)
+
+	sqlDB := sqlutils.MakeSQLRunner(tc.ServerConn(0 /* idx */))
+	sqlDB.CheckQueryResults(t,
+		"SELECT count(*) FROM [SHOW SCHEDULES FOR SQL STATISTICS]",
+		[][]string{{"0"}})
+
+	sqlDB.Exec(t, `SET CLUSTER SETTING version = $1`,
+		clusterversion.ByKey(clusterversion.SQLStatsCompactionScheduledJob).String())
+
+	// Change the recurrence cluster setting to force job monitor to create the
+	// sql stats compaction schedule.
+	sqlDB.Exec(t, "SET CLUSTER SETTING sql.stats.cleanup.recurrence = '@daily'")
+
+	// Wait for the change to be picked up by the job monitor.
+	sqlDB.CheckQueryResultsRetry(t,
+		"SELECT recurrence FROM [SHOW SCHEDULES FOR SQL STATISTICS]",
+		[][]string{{"@daily"}})
+
+	sqlDB.CheckQueryResults(t,
+		"SELECT count(*) FROM  [SHOW SCHEDULES FOR SQL STATISTICS]",
+		[][]string{{"1"}})
+}

--- a/pkg/sql/sqlstats/persistedsqlstats/scheduled_job_monitor.go
+++ b/pkg/sql/sqlstats/persistedsqlstats/scheduled_job_monitor.go
@@ -86,6 +86,10 @@ func (j *jobMonitor) start(ctx context.Context, stopper *stop.Stopper) {
 
 func (j *jobMonitor) registerClusterSettingHook() {
 	SQLStatsCleanupRecurrence.SetOnChange(&j.st.SV, func(ctx context.Context) {
+		if !j.isVersionCompatible(ctx) {
+			return
+		}
+		j.ensureSchedule(ctx)
 		if err := j.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
 			sj, err := j.getSchedule(ctx, txn)
 			if err != nil {
@@ -137,9 +141,9 @@ func (j *jobMonitor) getSchedule(
 }
 
 func (j *jobMonitor) ensureSchedule(ctx context.Context) {
-	clusterVersion := j.st.Version.ActiveVersionOrEmpty(ctx)
-	if !clusterVersion.IsActive(clusterversion.SQLStatsCompactionScheduledJob) {
-		log.Warningf(ctx, "cannot create sql stats scheduled compaction job because current cluster version is too low")
+	if !j.isVersionCompatible(ctx) {
+		log.Infof(ctx, "cannot create sql stats scheduled compaction job because current cluster version is too low")
+		return
 	}
 
 	var sj *jobs.ScheduledJob
@@ -166,6 +170,11 @@ func (j *jobMonitor) ensureSchedule(ctx context.Context) {
 	if err = CheckScheduleAnomaly(sj); err != nil {
 		log.Warningf(ctx, "schedule anomaly detected: %s", err)
 	}
+}
+
+func (j *jobMonitor) isVersionCompatible(ctx context.Context) bool {
+	clusterVersion := j.st.Version.ActiveVersionOrEmpty(ctx)
+	return clusterVersion.IsActive(clusterversion.SQLStatsCompactionScheduledJob)
 }
 
 // CheckScheduleAnomaly checks a given schedule to see if it is either paused

